### DISCUSSION TITLE
Any interest at all to make RTL_433 keys properly auto-discoverable?

### DISCUSSION THREAD

**Discussion by DigiH** [merbanan/rtl_433#3103] [Ideas]
With the current state of RTL_433 keys indiscriminately covering different types with the same named keys, it is hard, if not impossible, to correctly auto-discover such key correctly. 

The major example is the current battery_ok key, covering binary integer states for some devices, while others broadcast a float which can be nicely multiplied for a 0-100% battery level state.

Coming from the OpenMQTTGateway side, which uses the RTL_433_ESP port, this previously took over the RTL_433's own rtl_433_mqtt_hass.py script binary state conversion. This always was a thorn in my side side and a very ugly crutch tin my view, jumping from 100% to a nonsensical 0% only for any binary battery devices, making even reviewers think it 's a bug or what not.

I have mentioned this before in

https://github.com/merbanan/rtl_433/issues/2948

and eventually implemented a manual filtering in OpenMQTTGateway to correct this. With the recent update of RTL_433_ESP, with new decoders included,  this would mean going through all the added decoders again manually to see which new devices might need to be added to the manual filter at

https://github.com/1technophile/OpenMQTTGateway/blob/development/main/ZgatewayRTL_433.ino#L160

For which I do not have the time, nor the inclination at the moment.

It all boils down to the question if RTL_433 is willing and interested in making discoverable type distinctions for **battery_ok**, and possibly other similar keys, and owning up to the fact that it's 2024, well 2025 soon , and most users are non-technical relying on a preferable nicely well implemented auto-discovery, while renaming the float key to **battery_ok_f** and keeping **battery_ok** (for int/binary) would also not majorly break it for existing non-discovery users, even though **battery_level** and **battery_ok** would be more fitting keys.

Both the rtl_433_mqtt_hass.py script, as well as new devices in RTL_433_ESP would greatly benefit from this, presenting correctly battery device classes for auto-discovery, instead of the current state.

Thanks for reading my rant.

**Comment by gdt** [comment]
My take is that we should define semantics and then just change things as a bug fix, and not worry about compat with the previous world.

Secondarily, I lean to looking at what Home Assistant norms are, and esphome, and aligning to that.

I could also see battery_voltage for voltage in volts, separate from scaling to battery_level.  But I could also see the argument that this is duplicative and thus bad.

**Reply by zuckschwerdt** [reply]
There are already `battery_mV` and `battery_V` if there is some native data, so there is precedent :)

**Comment by zuckschwerdt** [comment]
There is always room for improvement. We don't like to break things for existing users, getting to `battery_ok` was a multi-year effort to replace all the different `battery` and `battery_low` fields with all kinds of values.

Adding such a central field, say `battery_level`, will need careful planing and definition -- we don't want to soon amend or replace it.

Whenever possible a field suffix should standardize the values, e.g. obvious suffixes like `_C`, `_F`, `kPa`, or `_km_h`.
Optimally an new field would use a well defined suffix -- `_level`, `_gauge`, `_pct`?

Some additional questions come to mind: what would be the defined top level? 100% i.e. 1.0? Should we cap there? What about values derived from battery mV with very fresh batteries above nominal voltage? What would be the nominal lowest values? 0% i.e. 0.0? Or should the nominal lowest value, e.g. derived from batteries or a battery_low flag be some random other value, 10%? Would there be a defined step for battery_warning, 30%? What about calculations that place minimum nominal battery voltage at, say, 10%, how much lower are they allowed to go? Should there be a cap or possibly negative values?

In summary, if we do something like this then we need to think of and document all the weird edge cases, thoroughly.

**Reply by DigiH** [reply]
While I agree that there might be pitfalls for other keys, which need proper investigations first, I think the battery_ok float or int is a pretty clear cut case, well suited as an initial case to bring this into the new auto-discovery world.

AFAIK all the battery_ok float values are actually received from the device broadcasts with values between 0.0 - 1.0, but if there really are some decoder derived from the voltage level, yes, I agree, they should be capped at 1.0, but this would already be the current case, where all battery_ok keys are presented as percentage values.

**Reply by zuckschwerdt** [reply]
We really have many decoders, at a quick glance at least these would need to be considered and possibly adapted to a new scheme:
- https://github.com/merbanan/rtl_433/blob/master/src/devices/fineoffset_ws90.c#99
- https://github.com/merbanan/rtl_433/blob/master/src/devices/govee.c#L194
- https://github.com/merbanan/rtl_433/blob/master/src/devices/govee.c#L391
- https://github.com/merbanan/rtl_433/blob/master/src/devices/radiohead_ask.c#L217 (missing battery_ok as there is no indication on nominal values)
- https://github.com/merbanan/rtl_433/blob/master/src/devices/fineoffset_ws80.c#L82
- https://github.com/merbanan/rtl_433/blob/master/src/devices/ambientweather_wh31e.c#L295
- https://github.com/merbanan/rtl_433/blob/master/src/devices/fineoffset.c#L636
- https://github.com/merbanan/rtl_433/blob/master/src/devices/steelmate.c#L66 (missing battery_ok, but there should be threshold voltages known)
- https://github.com/merbanan/rtl_433/blob/master/src/devices/fineoffset_wn34.c#L107

**Reply by DigiH** [reply]
> We really have many decoders, at a quick glance at least these would need to be considered and possibly adapted to a new scheme:

Have a look at my above linked OpenMQTTGateway fix, at least listing all the relevant float requiring decoder at the time of the RTL_433_ESP inclusion:

`            if (strcmp(pdevice->modelName, "Govee-Water") == 0 || strcmp(pdevice->modelName, "Govee-Contact") == 0 || strcmp(pdevice->modelName, "Archos-TBH") == 0 || strcmp(pdevice->modelName, "FineOffset-WH31L") == 0 || strcmp(pdevice->modelName, "Fineoffset-WH45") == 0 || strcmp(pdevice->modelName, "Fineoffset-WN34") == 0 || strcmp(pdevice->modelName, "Fineoffset-WS80") == 0 || strcmp(pdevice->modelName, "Fineoffset-WH0290") == 0 || strcmp(pdevice->modelName, "Fineoffset-WH51") == 0 || strcmp(pdevice->modelName, "Kedsum-TH") == 0 || strcmp(pdevice->modelName, "AVE") == 0 || strcmp(pdevice->modelName, "TPMS") == 0) {`


As I said, this could go on forever as a mainatnace nightmare, so a key naming fix is definitely the preferred solution, and would also make it possible for the rtl_433_mqtt_hass.py script to easily create correct auto-discovery entries.

**Reply by DigiH** [reply]
I also chose to make the float battery_ok decoder the exception, as the majority do actually have the binary battery_ok state, making this majority actually wrongly discovered at the moment with the rtl_433_mqtt_hass.py script.

**Comment by gdt** [comment]
So I guess the OP should say concretely what is wrong with `battery_ok`.  If we have  rule that it's 1/ok 0/low-or-not-happy, then we simply have bugs if a driver is not doing that.
 
Home  Assistant, as far as I understand it, uses "Battery Level" with 100% as full/new.    I think it should just be scaled to nominal, with 0% = 0V.    So basically battery_V / battery_nominal_V.    That leaves out a lot of complexity, which is good and bad.

We could also add battery_remaining to be intended to be 0% to 100% of the runtime from new batteries.  But that seems way harder than anybody is really going to do.  Plus, to do it right, you need to know about the battery that was inserted (alkaline vs NiMH, as the biggest example).

Personally, I want to know the voltage, but I like the 100%=fresh translation in HA.

But, agreed this is harder than I made it out to be and that a deliberate approach is in order.

**Reply by DigiH** [reply]
> So I guess the OP should say concretely what is wrong with `battery_ok`.

HA auto discovery has two distinct device classes for battery values, a sensor class, for numerical percentage values between 0%-100%, and a binary_switch device class, with battery Normal or battery low. For which I made the fix in OpenMQTTGateway for the then included RTL_433_ESP applicable devices. I can post the images again here to visualise the differences now with OMG, with I already liked to above

Correctly discovered float battery level

![image](https://github.com/user-attachments/assets/292257c2-91c3-4f48-9020-49055efa24ed)

Correctly discovered binary battery states

![image](https://github.com/user-attachments/assets/0ee2c874-3c4e-4275-90e8-69cd8f789c0e)

![image](https://github.com/user-attachments/assets/a142808a-4c9d-4617-b5d4-1b6fcedc8419)

Currently RTL_433 doesn't differentiate between these two types, float and binary, packing them all under a universal battery_ok key, making it impossible to properly discover this key, and resulting in the ugly crutch of converting the binary 0/1 state to either only o% or 100%, whereas a nice battery binary device class is available with HA discover.

**Reply by zuckschwerdt** [reply]
Basically: the below definition of `battery_ok` in https://triq.org/rtl_433/DATA_FORMAT.html#common-device-data is not well suited for data consumers.
> Battery status indication as a level between 0 (empty) and 1 (full). If the sensor can only report a binary status the value shall be 1 for "OK" and 0 for "LOW".

Proposed change: restrict `battery_ok` to 0/1 and introduce some field with float data.

Introducing a new field should come with a good definion. Possibly with a well defined suffix. It is a possibilty to improve which we might not easily get later on (breaking change).

**Reply by DigiH** [reply]
> Proposed change: restrict `battery_ok` to 0/1 and introduce some field with float data.

Totally agree with that :)

> Introducing a new field should come with a good definion. Possibly with a well defined suffix. It is a possibilty to improve which we might not easily get later on (breaking change).

If it's **battery_level**, **battery_percentage**, **battery_percentage_level** or whatever … that is up to the RTL_433 team, as long as it is distinguishable from the binary battery_ok, making it easily possible for OpenMQTTGateway, the rtl_433_mqtt_hass.py script or any other RTL_433 auto-discovery implementation to assign the correct device classes.

**Comment by gdt** [comment]
Sort of related:  https://andrew-codechimp.github.io/HA-Battery-Notes/

**Reply by DigiH** [reply]
Exactly my point, as described above and in the original post - to be able to correctly auto-discover the different battery_ok types.

Currently everything is forced not very nicely into a percentage value.

Also the links to the two different device classes for battery

https://www.home-assistant.io/integrations/sensor/#device-class

for the percentage level, and

https://www.home-assistant.io/integrations/binary_sensor/#device-class

for the binary battery_ok

**Comment by rct** [comment]
There is somewhat of a related issue of not really having boolean types in rtl_433 data schema. Using ints (or floats in battery_ok) allows for overloading which can be very tempting to overload in order to shoehorn something in. But in the long term this can be detrimental for having clear semantics for consumers. 

Other places this comes up is for most motion, security, leak, etc. sensors that using Home Assistant's terminology are naturally binary sensors.

(Side note: I think Home Assistant's binary_sensor device class for battery status didn't exist when the MQTT auto discovery script was created.)

----

Separate discussion, the current MQTT auto discovery script is somewhat of a dead end because it only works on the level of the individual data elements, so it can't make per model/device decisions. It also doesn't really keep state so hard to filter out false positive decodes and control when things get added. 

@DigiH - I'm not familiar with OpenMQTTGateway, but if it has support for the rtl_433 ESP port, could it be used to read from rtl_433 directly and replace the auto discovery script and/or mqtt gateway script for Home Assistant users?

**Reply by DigiH** [reply]
> @DigiH - I'm not familiar with OpenMQTTGateway, but if it has support for the rtl_433 ESP port, could it be used to read from rtl_433 directly and replace the auto discovery script and/or mqtt gateway script for Home Assistant users?

As OpenMQTTGateway runs on ESP32s it is thanks to @NorthernMan54's great rtl_433_ESP port which enabled it on the ESP32 devices with OpenMQTTGateway, either with a CC1101 attached, or the all-in-one LilyGo/Heltecs with their additional OLED screens, as well as publishing to MQTT.

![](https://user-images.githubusercontent.com/19808920/212491102-3ffd719b-9cb5-4494-b74f-49e7947c6887.png)

However, the sister app [Theengs Gateway](https://github.com/theengs/gateway), running on Windows, Linus and also available as a HA Add-on would be a candidate for such an integration, and there has already been a discussion started about the possibility, with me mentioning and linking to this however, stating that any such work really should wait until any key ambiguation has been resolved for proper auto-discovery device class assignments.

https://github.com/theengs/gateway/discussions/216

I mainly concentrate on Bluetooth decoders and their discoveries, but being a sucker for proper and nice auto-discoveries in general, I have also meddled quite a bit in the OMG rtl_433_ESP discovery side - as well as keeping to bring up this issue here 😉

**Reply by rct** [reply]
Thanks I couldn't tell whether OpenMQTTGateway was more of an app or library.   Either way the main idea is to avoid duplication of effort in terms of device discovery and maintaining a library of device details.

And thanks for that photo, I do need to get a couple of those ESP32s with RF modules.  It would be cool to see if those could be used to solve some reception issues.

**Reply by DigiH** [reply]
> I do need to get a couple of those ESP32s with RF modules.

Just make sure you get the compatible version ones, like the LilyGo LoRa32 V2.1, and the appropriate MHz version for your region/devices, e. g.

https://de.aliexpress.com/item/1005007766849442.html

and see if your devices' decoders have been ported to rtl_433_ESP

https://github.com/NorthernMan54/rtl_433_ESP?tab=readme-ov-file#enabled-device-decoders-from-rtl_433-release-2410

**Reply by rct** [reply]
@DigiH - I finally got around to getting one of the LilyGo Lora v2.1 boards and getting the OpenMQTTGateway firmware installed this afternoon. I wound up buying directly from LilyGo. There were some weird things going on with the AliExpress listings. 

Wow, I'm pretty impressed with how much this is doing for such little cost and effort!   All of those involved have done quite a nice job.  

I'll have to dig a bunch to see how viable a replacement it is for an RTL-SDR + RPI. It heard/discovered about 16 of my devices so far, so maybe 2/3.  Possibly 2 or 3 of these could cover the whole house.

Most notable thing missing is some of the messages from my Acurite Atlas weather station.  For some reason, it seems to mostly miss message type 37 (temperature and humidity). There might be something else going on there. 

So far I haven't figured out how to tell what it thinks the noise level looks like.

**Reply by DigiH** [reply]
Good to hear you are enjoying it @rct :) 

The main work has been done by @NorthernMan54 by porting the rtl_433 decoders and functionality to his rtl_433_ESP library, and @1technophile for the OpenMQTTGateway project. I mainly concentrate on the BLE decoding side and its discoveries, and through that also did some work on rtl_433 discovery, where I found the type ambiguity of identically named keys the biggest problem for producing correct discoveries - hence starting this discussion ;)

I also only have several LilyGos now around the house for my rtl_433 devices. Not so much for reception range issues, but because OOK and FSK decoders require different separate binaries to be installed and 433MHz and 868MHz decoders required different appropriate LilyGo hardware. This might also be an issue for part of your non-received devices. My RTL-SDR is resting in a drawer since then.

**Reply by rct** [reply]
There is a lot I need to dig into: improving reception, logging, and understanding the strengths/weaknesses of the hardware and platform. I see there is no timestamp in the message so correlating events between different receivers is gong to be problematic since it would be messaged from the time an MQTT client managed to get it from the broker with an unknown amount of latency. Clearly doesn't matter much for environment sensors like temperature, but gets more important for security sensors. Consider a door sensor that opens/closes or closes and opens again quickly. How do you know the last MQTT message you got is the correct state?  It could be from a marginal receiver overwriting the correct state from a better performing receiver. 

I also need to figure out why my device gets into a mode where it reboots frequently.

![image](https://github.com/user-attachments/assets/cfa01121-188d-4c76-b8c8-80bb0f935c85)

I'll follow up in the appropriate repo once I have some clues. Sure would be nice if it was syslogging. Guess I might have to connect it to a computer so I can see the serial log output

**Reply by DigiH** [reply]
> I see there is no timestamp in the message …

Might this be helpful? Not available as a runtime settable MQTT command (so far), but as a custom environment/build build_flag

https://github.com/1technophile/OpenMQTTGateway/blob/development/main/User_config.h#L410-L412

with an example MQTT message of

{"model":"LaCrosse-TX141THBv2","id":37,"channel":0,"battery_ok":1,"temperature_C":-22.9,"humidity":78,"test":"No","mic":"CRC","protocol":"LaCrosse TX141-Bv2, TX141TH-Bv2, TX141-Bv3, TX141W, TX145wsdth, (TFA, ORIA) sensor","rssi":-50,"duration":175992,**"UTCtime":"2024-12-30T19:14:43Z"**}

or the UNIX timestamp underneath.

**Reply by DigiH** [reply]
> I also need to figure out why my device gets into a mode where it reboots frequently.

I usually find this is happening when other devices, which I do not own, pass by my house, but being in a very rural area this doesn't happen very often for me, once every few days or weeks, definitely not as frequently as in your uptime screenshot above.

@NorthernMan54 will be better suited to answer this, but he also provisioned for something like this in his library with the [`MY_DEVICES`](https://github.com/NorthernMan54/rtl_433_ESP?tab=readme-ov-file#compile-definition-options) flag, to only try and decode a specific set of decodable devices.

**Comment by gdt** [comment]
Doctrine questions:
  - if we define multiple things, that various data consumers  might want, should drivers synthesize things they don't have natively have, or should they simply try to translate and omit things they don't know
  - should we have a "battery_level", defined as voltage / nominal, so that 1.0 is nominally full (I say yes)
    + if so, should we clamp at 1, or not clamp (I say not clamp)
    + or should it be 100* v/nominal, so it's percent (fine with me, doesn't really matter, pick one, go with % to align with HA)
  - should we have a "battery_percent", defined as 0% is used up and 100% is new, and map a device's estimate of that, if it exists to battery_percent?  (I say yes, but don't make things up and don't pressure drivers to have this.)    An example is WH41 that has 0%-100% in 20% steps.
  - should we try to synthesize battery_percent from battery_level, knowing how the batttery works (I say very very maybe, feels like trouble)
  - should we simply fix all drivers that emit battery_ok as other than binary to emit battery_level or _percent instead (I say yes)

**Reply by rct** [reply]
I don't think rtl_433 should synthesize anything.  If all the device gives is a binary battery low or battery ok, I'd like to know in the consumer and/or display that there will only ever be 2 values. I'd rather not have to think about why that battery is still at 100% when I not sure how many months have elapsed since it was last changed. 

BTW, Interesting find on the HA-Battery-Notes integration.  I have to take a serious look at that.  I always thought at a simple/useful thing Home Assistant could do is add a markdown card to the device info page that you could edit to store battery notes, link to manuals, etc.  I like Zigbee2MQTTs access to doc links.

**Reply by zuckschwerdt** [reply]
Re boolean types: this can be handled much nicer, type-safe and extensible, now that #2667 has landed. There is already one example with `data_hex()`, see last comment on the PR. Much easier to reason about decoder outputs then -- scriptable and thus likely introspectable.

My opinion on the open questions:
- Seeing percentage support in consumers like HA I favor `battery_pct` instead of `battery_level`
- I don't want to clamp, it obscures interesting edge-case data
- I'd say 0% target is okay as lower value, if applicable something like `100 * (v_curr - v_min) / (v_max - v_min)`
- as proposed move all float `battery_ok` fields to `battery_pct`, have `battery_ok` as bool.
- unsure on the raw voltage or `battery_pct` mapping to a synthetic `battery_ok` though?

**Reply by gdt** [reply]
I guess there is enough hesitation about synthetic that for now we should say no, and we can add it later.  It's harder to go the other way.

Let's spell out battery_percent if we use that.

I am not a fan of `v_min`.  There's no way to figure it out.  It makes interpretation hard.

Trying to converge to something we can all be ok with,  I lean to two keys, depending on what the device does:
  - battery_level: simple scaled battery voltage divided by nominal, no clamping, for devices that read voltage
  - battery_percent: 0% dead to 100% new for devices that give a percentage of capacity remaining (again no clamping if they report 105% for a super-good battery)

we can then figure out how HA should take level and make percent, but at least we'll be reporting clean, useful data.

**Reply by zuckschwerdt** [reply]
I see `_pct` as a general (reusable) unit suffix.

We currently have some instances where a battery level (sometimes derived from a voltage) is output as float in `battery_ok`, my understanding was to make `battery_ok` a bool (how?) and output the current float x100 as `battery_pct`.
The formula would be/should be something like `100 * (v_curr - v_min) / (v_max - v_min)` -- some `v_min` is needed to cover the range down to 0%, right? Some conservative guess on `v_min` to usually get somewhat close to 0% maybe? But that's just as a guideline, decoders already have some scaling present.
 
For other scalings, e.g. the 5-bar example we would need to decide if it's 1 to 5 x20, so 20% as lowest, or 0 to 4 x25.

I'm not sure what `battery_level` would be? It looks like `battery_pct` but different scaling?

**Comment by gdt** [comment]
discovery probably needs to be upleveled, perhaps to some descriptor emitted by the driver, or in rtl_433 more properly.   That's a much bigger rototill.  Otherwise we're doing it twice.

**Reply by rct** [reply]
Another motivation for upleveling/modernization is from the 2024.11 release notes

> added support for [MQTT device-based auto discovery](https://www.home-assistant.io/integrations/mqtt/#device-discovery-payload). This allows MQTT devices to be set up and discovered once instead of separately for each entity; which is much more efficient. Nice improvement!

This makes things more device-centric where device info / keys like availability, origin, state_topic can all be shared.  You wind up with basically one JSON config doc per device instead of one for each entity.  The amount of redundant/verbose config info can be greatly reduced.

Also with this style of configuration it would make it preferable to have rtl_433 publish an MQTT topic per device instead of a separate topic for each data item.  Use a value_template to pull the desired entities out of the JSON state.   It would cut the MQTT message load down considerably.  (Yes you could do a lot of that now, but it is a lot more clumsy and difficult to maintain.)

Note: the Home Assistant MQTT docs are more confusing than they need to be because they are using a lot of abbreviations.

**Reply by 1technophile** [reply]
> discovery probably needs to be upleveled, perhaps to some descriptor emitted by the driver, or in rtl_433 more properly. That's a much bigger rototill. Otherwise we're doing it twice.

One approach we've explored for BLE devices involves placing the device properties descriptor at the device level. If you're curious to see how this works in practice, you can try it out at https://parser.theengs.io.
With the decoder library (https://decoder.theengs.io), it can not only generate measurement values from undecoded messages but also provide device descriptions and properties. We found this particularly useful for auto-discovery integration, though there are certainly other valid approaches out there. Feel free to check it out if it aligns with your needs.

**Comment by gdt** [comment]
There are two ways to represent information that is derived from voltage.  Well, 3.  One is to just output the voltage, and I would say any decoder that reads voltage should output battery_V (not mV, but V, this is using SI units vs some synthetic -- we already convert to metric units otherwise).

Then, there is "battery_level", which is battery voltage divided by nominal battery voltage * 100%,    That's a way to have something that behaves exactly like voltage but is on a common scale if graphing different kinds of devices.    For example I am monitoring (not with rtl_433) 3 UPS units.  2 have 2x 12V lead acid, and 1 has 1.  It would be nice to look at that as /24V and /12V since then it lines up nicely.   Note that this has no belief about remaining life and curve shapes - it is not obscuring the underlying data, just fixed scaling.

Finally there is battery_pct (or _percent), which is an estimate of remaining life, from 100% at new battery to 0% at  failure and in theory linear in time.  This is very tricky.    Some devices report things that are like this and mapping is straightforward. Making up percent life from voltage is I think aspiratiional rather than reliable, absent device-specific calibration.

If you think there is some min, and want to synthesize percent from new-battery-min (say 1.6V to 1.3V for AA, but that fails for NiMH), then I don't have a problem with that, but I would like to see battery_level which is 100% for 1.5V, sort of "percent of battery nominal voltage".  As in also see, not instead.

**Comment by DigiH** [comment]
Another ambiguity and inconsistency is with the **uv** and **uvi**  keys.

The Vevor 7-in-1 has the UV key actually being the **UV Index**, while my own Fineoffset-WH65B has both UV and UVI, which we discover as **UV Level/Raw** (UV) and **UV Index** (UVI), already causing a **UV Index** inconsistency and discovery mayhem with these two decoders without having even looked at any other UV and UVI implementations.

**Comment by zuckschwerdt** [comment]
@DigiH yes, all UV Index fields should be named "uvi", please review #3131

**Reply by DigiH** [reply]
Having a look at #3131 I wonder what the two **Bresser**, the **Cotech** and **FineOffset WS80** one decimal doubles are

> DATA_FORMAT, "%.1f", DATA_DOUBLE

Either some very obscure way of presenting a usually integer UV Index with additional in between decimals, or also some other raw data not indicating the UV Index at all?!? With the latter also requiring to rename their "UV Index" label.

The initial implementations of these decoders might have moire insight into this, or if someone has one of these stations.
